### PR TITLE
server: a better way to handle killed connection

### LINKED ISSUE #24031: Killing connections needs cooperation from the to-be-killed connection
### 1. Minimal reproduce step (Required)

- Run TiDB playground.
- Connect two sessions with `mysql --host 127.0.0.1 --port 4000 -u root -p`
- From connection A: run `SHOW PROCESSLIST` and get the connection ID of the other connection (`mysql` reports its own connection id when starting up)
- Now from connection A issue: `KILL TIDB CONNECTION <id>` where ID is the connection ID of the other connection.
- Now run `SHOW PROCESSLIST` again.

Note that this is with a single `tidb` node.

### 2. What did you expect to see? (Required)

The killed connection gone from the processlist.

### 3. What did you see instead (Required)

The other connection still being in the processlist until it tries to run a query.

Once a query like `do 1` is attempted on connection B the connection is gone from the processlist and connection B shows `ERROR 2013 (HY000): Lost connection to MySQL server during query`

### 4. What is your TiDB version? (Required)

```
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v5.0.0
Edition: Community
Git Commit Hash: bdac0885cd11bdf571aad9353bfc24e13554b91c
Git Branch: heads/refs/tags/v5.0.0
UTC Build Time: 2021-04-06 16:36:29
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve?

Issue Number: close #24031, this PR also reverts #29212.

Problem Summary:

### What is changed and how it works?

The root cause of #24031 is that when a connection is idle, the goroutine is blocked at:

https://github.com/pingcap/tidb/blob/4a0d387e1ff1b508bbb60d484d97e4ac2a5ef2c7/server/conn.go#L1068

And the stack:

```
#	0x13fc361	bufio.(*Reader).Read+0x221					/home/bb7133/Softwares/go/src/bufio/bufio.go:227
#	0x34c8fba	github.com/pingcap/tidb/server.bufferedReadConn.Read+0x5a	/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/buffered_read_conn.go:31
#	0x1342886	io.ReadAtLeast+0x86						/home/bb7133/Softwares/go/src/io/io.go:328
#	0x34a96e4	io.ReadFull+0x84						/home/bb7133/Softwares/go/src/io/io.go:347
#	0x34a96ab	github.com/pingcap/tidb/server.(*packetIO).readOnePacket+0x4b	/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/packetio.go:86
#	0x34a9aee	github.com/pingcap/tidb/server.(*packetIO).readPacket+0x4e	/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/packetio.go:117
#	0x3479624	github.com/pingcap/tidb/server.(*clientConn).readPacket+0x1e4	/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/conn.go:397
#	0x34795ea	github.com/pingcap/tidb/server.(*clientConn).Run+0x1aa		/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/conn.go:1068
#	0x34b2d1d	github.com/pingcap/tidb/server.(*Server).onConn+0x12bd		/home/bb7133/Projects/gopath/src/github.com/pingcap/tidb/server/server.go:554
```

Because of that, the goroutine is not able to deal with the `KILLED` flag,  release the resource it is holding and stop itself immediately.

In order to solve that, we need to make `conn.Read()` *interruptable* but there is no straightforward way in Go to do that. Some references:

1) https://github.com/golang/go/issues/20280: a lot of discussions/arguments without a clear conclusion.
2) [Canceling I/O in Go Cap’n Proto](https://medium.com/@zombiezen/canceling-i-o-in-go-capn-proto-5ae8c09c5b29): mentioned in `go/issues/20280`
3) https://github.com/google/mtail/pull/497: a context-based implementation for canceling the `Read()`

For the approach introduced in 2 and 3, they are generally the same as this PR: setting `SetReadDeadline` in another goroutine. I cannot find any material describing if doing so is thread-safe, so it should be implementation-dependent and might not be safe, but it might not be a real problem considering we're about the kill the connection and the read buffer/status will be abandoned.

### Alternatives

* `SHOW PROCESSLIST` (and infoschema) is modified to show the State as `Killed`, as mentioned by @morgo in https://github.com/pingcap/tidb/issues/24031#issuecomment-893095960, the result of `SHOW PROCESSLIST` can be clear to the user but it doesn't solve the delayed 'release lock' issue(see `Case 2` in 'Manual test' part).
* Instead of setting the read timeout to `waitTimeout`, the code is instead modified to have a hard coded `2s` timeout, but loops for up to `waitTimeout` retrying a read..., also mentioned by @morgo in https://github.com/pingcap/tidb/issues/24031#issuecomment-893095960, the potential thread-safe concern can be avoided but we still have at most `2s` delay for killing an idle connection and the code would be complicated.
* Instead of setting `SetReadDeadline()`, `bufReadConn.Close()` can be another solution. It is basically the same with `SetReadDeadline()` IMHO.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Case1:

```
Session1> (in idle state, with PROCESS_ID=3)
Session2> KILL TIDB 3;
Session2> SHOW PROCESSLIST; (Can be confirmed that Session1 is killed)
```

Case2:

```
Session1> CREATE TABLE t1(a INT);
Session1> INSERT INTO t1 values (1);
Session1> BEGIN PESSIMISTIC;
Session1> SELECT * FROM t1 WHERE a=1 FOR UPDATE;
Session1> (in idle state, with PROCESS_ID=3)
Session2> BEGIN PESSIMISTIC;
Session2> SELECT * FROM t1 WHERE a=1 FOR UPDATE; (Session 2 is blocked and waiting for the lock)
Session3> KILL TIDB 3; (Can be confirmed that Session 1 is killed and Session2 is able to acquire the lock immediately).
```

Side effects

- None

### Release note

```release-note
fix the issue that `KILL TIDB` doesn't take effect immediately on idle connections
```

### CODE REVIEW

**Comment by bb7133** [comment]
PTAL @djshow832 @tiancaiamao @morgo @dveeden

File: server/conn.go [hunk 1]
```diff
@@ -1069,12 +1069,16 @@ func (cc *clientConn) Run(ctx context.Context) {
...
 			if terror.ErrorNotEqual(err, io.EOF) {
 				if netErr, isNetErr := errors.Cause(err).(net.Error); isNetErr && netErr.Timeout() {
-					idleTime := time.Since(start)
-					logutil.Logger(ctx).Info("read packet timeout, close this connection",
-						zap.Duration("idle", idleTime),
-						zap.Uint64("waitTimeout", waitTimeout),
-						zap.Error(err),
-					)
+					if atomic.LoadInt32(&cc.status) == connStatusWaitShutdown {
+						logutil.Logger(ctx).Info("read packet timeout because of killed connection")
```
Comment (jackysp) [inline server/conn.go:1073]: Maybe debug level is better? TiDB shut down when a large number of connections exist may cause too much logging?

**Review by jackysp — APPROVED** [review]
LGTM

**Review by djshow832 — APPROVED** [review]

**Comment by zimulala** [comment]
@bb7133 
It seems that we need to cherry-pick to v5.4, but we don't do it.

**Pull request merged**